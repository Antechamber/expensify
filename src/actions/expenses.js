import database from '../firebase/firebase'
import { ref, push, get, child, set } from 'firebase/database'

export const addExpense = (expense) => ({
  type: 'ADD_EXPENSE',
  expense
})

// this action 'constructor' first updates the firebase db, then dispatches the original addExpense function
export const startAddExpense = (expenseData = {}) => {
  return (dispatch) => {
    const {
      description = '',
      note = '',
      amount = 0,
      createdAt = 0
    } = expenseData
    const expense = { description, note, amount, createdAt }

    return push(ref(database, 'expenses'), expense).then((ref) => {
      dispatch(addExpense({
        id: ref.key,
        ...expense
      }))
    })
  }
}

export const removeExpense = ({ id } = {}) => ({
  type: 'REMOVE_EXPENSE',
  id
})

export const startRemoveExpense = ({ id } = {}) => {
  return (dispatch) => {
    return set(ref(database, `expenses/${id}`), null)
      .then(() => dispatch(removeExpense({ id })))
  }
}

export const editExpense = (id, updates) => ({
  type: 'EDIT_EXPENSE',
  id,
  updates
})

// set expenses in store from firebase db

export const setExpenses = (expenses = []) => ({
  type: 'SET_EXPENSES',
  expenses
})

export const startSetExpenses = () => {
  return (dispatch) => {
    return get(child(ref(database), 'expenses'))
      .then((snapshot) => {
        const expenses = []
        if (snapshot.exists()) {
          snapshot.forEach((childSnapshot) => {
            expenses.push({
              id: childSnapshot.key,
              ...childSnapshot.val()
            })
          })
        }
        dispatch(setExpenses(expenses))
      })
  }
}
