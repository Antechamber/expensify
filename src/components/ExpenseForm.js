import React from "react"
import moment from 'moment'
import 'react-dates/initialize';
import { SingleDatePicker } from "react-dates"

export default class ExpenseForm extends React.Component {
    constructor(props) {
        super(props)
        this.state = {
            description: props.expense ? props.expense.description : '',
            note: props.expense ? props.expense.note : '',
            amount: props.expense ? (props.expense.amount / 100).toString() : '',
            createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
            calendarIsFocused: false,
            error: ''
        }
    }

    onDescriptionChange = (e) => {
        const description = e.target.value
        this.setState(() => ({ description }))
    }
    onNoteChange = (e) => {
        const note = e.target.value
        this.setState(() => ({ note }))
    }
    onAmountChange = (e) => {
        const amount = e.target.value
        if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
            this.setState(() => ({ amount }))
        }
    }
    onDateChange = (createdAt) => {
        if (createdAt) {
            this.setState(() => ({ createdAt }))
        }
    }
    onCalendarFocusChange = ({ focused }) => {
        this.setState(() => ({ calendarIsFocused: focused })) // when focus changes, keep track in state
    }
    onSubmit = (e) => {
        e.preventDefault()
        if (!this.state.description || !this.state.amount) {
            this.setState(() => ({ error: 'Please provide description and amount' }))
        } else {
            this.setState(() => ({ error: undefined }))
            this.props.onSubmit({
                description: this.state.description,
                amount: parseFloat(this.state.amount, 10) * 100,
                createdAt: this.state.createdAt.valueOf(),
                note: this.state.note
            })
        }
    }

    render() {
        return (
            <div>
                {this.state.error && <p>{this.state.error}</p>}
                <form onSubmit={this.onSubmit}>
                    <input
                        name="description"
                        type="text"
                        placeholder="Description"
                        autoFocus
                        value={this.state.description}
                        onChange={this.onDescriptionChange}
                        autoComplete="off"
                    />
                    <input
                        name="amount"
                        type="text"
                        placeholder="Amount"
                        value={this.state.amount}
                        onChange={this.onAmountChange}
                        autoComplete="off"
                    />
                    <SingleDatePicker
                        date={this.state.createdAt} // momentPropTypes.momentObj or null
                        onDateChange={this.onDateChange} // PropTypes.func.isRequired
                        focused={this.state.calendarIsFocused} // PropTypes.bool
                        onFocusChange={this.onCalendarFocusChange} // PropTypes.func.isRequired
                        id="expenseFormSingleDatePicker" // PropTypes.string.isRequired,
                        numberOfMonths={1}
                        isOutsideRange={() => false}
                    />
                    <textarea
                        placeholder="Add a note (optional)"
                        value={this.state.note}
                        onChange={this.onNoteChange}
                    >
                    </textarea>
                    <button
                        onClick={null}
                    >Submit</button>
                </form>
            </div>
        )
    }
}