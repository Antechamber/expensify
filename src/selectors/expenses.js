// get visible expenses (i.e. use the filters object)
import moment from "moment"

export default (expenses, { text, sortBy, startDate, endDate }) => {
    return expenses.filter((expense) => {
        const createdAtMoment = moment(expense.createdAt)
        const filterMatch = typeof text !== 'string' || expense.description.toLowerCase().includes(text.toLowerCase())
        const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment, 'day') : true
        const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment, 'day') : true

        return startDateMatch && endDateMatch && filterMatch && startDateMatch && endDateMatch
    }).sort((a, b) => {
        if (sortBy === 'date') {
            return a.createdAt < b.createdAt ? 1 : -1
        } else if (sortBy === 'amount') {
            return a.amount < b.amount ? 1 : -1
        }
    })
}