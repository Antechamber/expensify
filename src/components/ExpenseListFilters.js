import 'react-dates/initialize'
import React, { useState } from "react"
import { connect } from "react-redux"
import { DateRangePicker } from "react-dates"
import { setTextFilter, sortByDate, sortByAmount, setEndDate, setStartDate } from '../actions/filters'

export const ExpenseListFilters = (props) => {
    // useState
    const [calendarFocused, setCalendarFocus] = useState(null)

    // event handlers
    const onDatesChange = ({ startDate, endDate }) => {
        props.setStartDate(startDate)
        props.setEndDate(endDate)
    }
    const onFocusChange = (calendarFocused) => {
        setCalendarFocus(calendarFocused)
    }
    const onTextChange = (e) => {
        props.setTextFilter(e.target.value)
    }
    const onSortChange = (e) => {
        e.target.value === 'date' ? props.sortByDate() : props.sortByAmount()
    }

    return (
        <div>
            <label htmlFor="textFilter">Filter</label>
            <input id="textFilter" type="text" value={props.filters.text} onChange={onTextChange} />
            <select
                value={props.filters.sortBy}
                onChange={onSortChange}>
                <option value="date">Date</option>
                <option value="amount">Amount</option>
            </select>
            <DateRangePicker
                startDate={props.filters.startDate} // momentPropTypes.momentObj or null,
                startDateId="Expense_Dashboard_Page_Filter_Start_Date" // PropTypes.string.isRequired,
                endDate={props.filters.endDate} // momentPropTypes.momentObj or null,
                endDateId="Expense_Dashboard_Page_Filter_End_Date" // PropTypes.string.isRequired,
                onDatesChange={onDatesChange} // PropTypes.func.isRequired,
                focusedInput={calendarFocused} // PropTypes.oneOf([START_DATE, END_DATE]) or null,
                onFocusChange={onFocusChange} // PropTypes.func.isRequired,
                showClearDates={true}
                numberOfMonths={1}
                isOutsideRange={() => false}
            />
        </div>
    )
}

const mapStateToProps = (state) => ({
    filters: state.filters
})

const mapDispatchToProps = (dispatch) => ({
    sortByDate: () => dispatch(sortByDate()),
    sortByAmount: () => dispatch(sortByAmount()),
    setStartDate: (startDate) => dispatch(setStartDate(startDate)),
    setEndDate: (endDate) => dispatch(setEndDate(endDate)),
    setTextFilter: (value) => dispatch(setTextFilter(value))
})

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters)