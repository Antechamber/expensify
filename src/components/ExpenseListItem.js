import React from "react";
import { connect } from 'react-redux'
import { removeExpense } from '../actions/expenses'
import { Link } from "react-router-dom";
import moment from 'moment'

const ExpenseListItem = ({ description, amount, createdAt, id }) => {
    return (
        <div>
            <Link to={`/edit/${id}`}>{description}</Link>
            <p>{amount}</p>
            <p>{moment(createdAt).format("dddd, MMMM Do YYYY")}</p>
        </div>
    )
}

export default ExpenseListItem