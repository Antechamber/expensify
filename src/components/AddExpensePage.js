import React from 'react'
import ExpenseForm from './ExpenseForm'
import { connect } from 'react-redux'
import { addExpense } from '../actions/expenses'
import { useNavigate } from 'react-router-dom'

export class AddExpensePage extends React.Component {
    onSubmit = (expense) => {
        this.props.addExpense(expense)
        this.props.navigate('/')
    }
    render() {
        return (
            <div>
                <h1>Add Expense</h1>
                <ExpenseForm
                    onSubmit={this.onSubmit}
                />
            </div>
        )
    }
}

// since hooks are invalid inside class component, I'm wrapping AddExpensePage
// in functional component to use useNavigate. 
// converted from frunctional to class based so that ExpenseForm would not have onSubmit defined in line.
// in line functions in functional components are recalculated every rerender. Should try to avoid in line declarations when possible
const WrappedAddExpensePage = (props) => {
    const navigate = useNavigate()
    return <AddExpensePage navigate={navigate} {...props} />
}

const mapDispatchToProps = (dispatch) => ({
    addExpense: (expense) => dispatch(addExpense(expense))
})

export default connect(undefined, mapDispatchToProps)(WrappedAddExpensePage)
