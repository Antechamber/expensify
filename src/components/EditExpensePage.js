import React from 'react'
import { useNavigate, useParams } from 'react-router-dom'
import { connect, useSelector } from 'react-redux'
import ExpenseForm from './ExpenseForm'
import { startEditExpense, startRemoveExpense } from '../actions/expenses'

export class EditExpensePage extends React.Component {
  editExpense = (expense) => {
    this.props.editExpense(this.props.expense.id, expense)
    this.props.navigate('/')
  }
  removeExpense = () => {
    this.props.removeExpense(this.props.expense.id)
    this.props.navigate('/')
  }
  render() {
    return (
      <div>
        <ExpenseForm
          expense={this.props.expense}
          onSubmit={this.editExpense}
        />
        <button onClick={this.removeExpense}>Remove</button>
      </div>
    )
  }


}

const EditExpensePageWrapped = (props) => {
  // get id out of url params
  const { id } = useParams()
  // useSelector has access to the state in the redux store.
  // using it here to find expense in store that matches the id above
  // then mapping that to props (hooks unavailable in mapStateToProps)
  const expense = useSelector((state) => {
    return state.expenses.find((expense) => expense.id === id)
  })
  const navigate = useNavigate()
  return (
    <EditExpensePage
      {...props}
      expense={expense}
      navigate={navigate}
    />
  )
}
// mapping the dispatch method to props of component.
// this abstracts the usage to testing is easier/disconnected from redux
const mapDispatchToProps = (dispatch) => ({
  editExpense: (id, expense) => dispatch(startEditExpense(id, expense)),
  removeExpense: (id) => dispatch(startRemoveExpense({ id }))
})


export default connect(undefined, mapDispatchToProps)(EditExpensePageWrapped)
