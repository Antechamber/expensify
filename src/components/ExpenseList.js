import React from 'react'
import { connect } from 'react-redux'
import ExpenseListItem from './ExpenseListItem'
import selectExpenses from '../selectors/expenses'

export const ExpenseList = (props) => (
    <div>
        <h1>Expense List</h1>
        {props.expenses.length === 0 ? (
            <p>No Expenses</p>
        ) :
            props.expenses.map((expense) => (
                <ExpenseListItem key={expense.id} {...expense} />
            ))

        }
        <br />
    </div>
)

// mapStateToProps configures which properties of the state are available to the wrapped component via props
const mapStateToProps = (state) => {
    return {
        expenses: selectExpenses(state.expenses, state.filters)
    }
}

// connect wraps a connection to the store around ExpenseList
export default connect(mapStateToProps)(ExpenseList)
