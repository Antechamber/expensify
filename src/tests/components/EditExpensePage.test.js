import React from "react"
import { shallow } from "enzyme"
import { EditExpensePage } from "../../components/EditExpensePage"
import expenses from '../fixtures/expenses'

let wrapper, editExpense, removeExpense, navigate, expense

beforeEach(() => {
    editExpense = jest.fn()
    removeExpense = jest.fn()
    navigate = jest.fn()
    expense = expenses[2]
    wrapper = shallow(
        <EditExpensePage
            editExpense={editExpense}
            removeExpense={removeExpense}
            expense={expense}
            navigate={navigate}
        />
    )
})

test('Should render EditExpensePage correctly', () => {
    expect(wrapper).toMatchSnapshot()
})

test('Should handle editExpense', () => {
    const updates = {
        ...expense,
        amount: 9999
    }
    wrapper.find('ExpenseForm').prop('onSubmit')(updates)
    expect(navigate).toHaveBeenLastCalledWith('/')
    expect(editExpense).toHaveBeenLastCalledWith(expense.id, updates)
})

test('Should handle removeExpense', () => {
    wrapper.find('button').simulate('click')
    expect(navigate).toHaveBeenLastCalledWith('/')
    expect(removeExpense).toHaveBeenLastCalledWith(expense.id)
})