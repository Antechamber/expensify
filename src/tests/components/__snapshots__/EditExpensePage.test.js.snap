// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render EditExpensePage correctly 1`] = `
<div>
  <ExpenseForm
    expense={
      Object {
        "amount": 799,
        "createdAt": 0,
        "description": "Flour",
        "id": "3",
        "note": "Makin' that bread",
      }
    }
    onSubmit={[Function]}
  />
  <button
    onClick={[Function]}
  >
    Remove
  </button>
</div>
`;
